File: string/1585. Check If String Is Transformable With Substring Sort Operations/js/main.js
// Time O(N)
// Space O(N)
const isTransformable = (s, t) => {
  let map = new Map();
  let n = s.length;

  for (let i = n; i >= 0; i--) {
    let x = parseInt(s[i]);
    if (!map.has(x)) map.set(x, []);
    map.get(x).push(i);
  }

  for (let i = 0; i < n; i++) {
    let num = parseInt(t[i]);

    if (!map.has(num)) {
      return false;
    }

    for (let prevNum = num - 1; prevNum >= 0; prevNum--) {
      if (!map.has(prevNum)) continue;

      if (map.get(prevNum).length && last(map.get(prevNum)) < last(map.get(num))) {
        return false;
      }
    }

    map.get(num).pop();
  }

  return true;

  function last(x) {
    return x[x.length - 1];
  }
};

File: string/1585. Check If String Is Transformable With Substring Sort Operations/py/main.py
import collections

class Solution:
    def isTransformable(self, s, t):
      countMap = collections.defaultdict(list)

      for i in reversed(range(len(s))):
        key = int(s[i])
        countMap[key].append(i)

      for i in range(len(t)):
        num = int(t[i])

        if num not in countMap:
          return False

        for prevNum in reversed(range(num)):
          if prevNum not in countMap:
            continue
          
          if len(countMap[prevNum]) and countMap[prevNum][-1] < countMap[num][-1]:
            return False

        countMap[num].pop()
      return True
